File: src/mujoco_skills/orchestrator/plan_edits.py
"""Backend application of manual plan-edit deltas (compound_turn_integration_
spec.md D2b, revising D2 / §5 item 16).

The frontend's `planEdits.ts` mutations are the browser-side source of truth
while a Gantt/scene edit is in progress; ScenePage keeps the PENDING batch
(not yet applied) and sends it with whichever turn applies it. D2's original
design had the backend replay that batch onto a freshly decomposed authored
plan every turn (`decompose(semanticActions)`) -- correct for a plain author
turn, but wrong for an edit-tail sync: the user edits the RESOLVED plan (D1's
only user-visible plan), which contains resolver artifacts (`go_to_rest`
departure legs, cross-robot `after` edges) a fresh decompose cannot express.
A delta targeting one of those had no target under D2 -- dropped with no
effect -- while the resolved plan's prior repairs were discarded in the same
stroke by the rebuild-from-scratch itself. D2b's fix: `stream_compound_turn`
picks a per-path plan (previous resolved plan for "edit"/"sync", the D1b graft
for a pure-append author turn, a clean recompute otherwise) and THIS module
applies the batch onto whichever one it picked -- see that function's
docstring for the full table. This module itself is plan-agnostic: it walks
whatever `tasks`/`steps` are in the dict it's given, resolver artifacts
included.

Deltas are targeted by ids that must survive onto the plan they are applied
to: a task-level edit targets the task/group id (stable as long as the action
itself is unmodified), a step-level edit targets the step id
(`f"{action_id}:s{index}"` for an authored step, or the resolver's own id for
an artifact like `f"{robot}#go_to_rest"`). When a target does not exist in
that plan -- the action was removed, a field change shifted its step shape,
or (D2b) the whole batch is being dropped because a non-pure-append author
turn replaced the plan out from under it -- the delta is dropped, never
silently: the caller must surface ``dropped_edit_text`` for every dropped
delta (a visible activity line), which is why `replay_edits` returns the
dropped list rather than swallowing it.
"""

from __future__ import annotations

import copy

from mujoco_skills.orchestrator.compiler_v2_plan import promote_edited_artifact


# Deltas that change WHO does what or in WHAT ORDER, as opposed to where a
# path runs. Mirrors the frontend's `draftEditImpact` "structural" bucket
# (ScenePage's editDraft call sites) -- the two lists must stay in step, but
# they are kept separate on purpose: the frontend's drives a button label,
# this one gates a plan rewrite, and a client is not trusted to decide that.
STRUCTURAL_OPS = frozenset({"remove_task", "move_task", "set_task_robot", "set_step_after"})


def semantic_removal_ids(edits: list[dict] | None) -> list[str]:
    """Return the ordered, deduplicated semantic task ids staged for removal."""
    result: list[str] = []
    for edit in edits or []:
        if (edit or {}).get("op") != "remove_task":
            continue
        action_id = ((edit or {}).get("target") or {}).get("actionId")
        if isinstance(action_id, str) and action_id and action_id not in result:
            result.append(action_id)
    return result


def without_semantic_removals(edits: list[dict] | None) -> list[dict]:
    """Deltas left for plan-level replay after semantic removals are applied."""
    return [edit for edit in edits or [] if (edit or {}).get("op") != "remove_task"]


def has_structural_edit(edits: list[dict] | None) -> bool:
    """True when this batch changes the coordination premises (allocation or
    ordering), which is what invalidates the resolver's departure anchors --
    see decompose.strip_stale_departures for why that matters."""
    return any((edit or {}).get("op") in STRUCTURAL_OPS for edit in edits or [])


def _find_step(plan: dict, step_id: str | None) -> dict | None:
    if not step_id:
        return None
    for task in plan.get("tasks", []):
        for step in task.get("steps", []):
            if step.get("id") == step_id:
                return step
    return None


def _known_step_ids(plan: dict) -> set[str]:
    return {
        step["id"]
        for task in plan.get("tasks", [])
        for step in task.get("steps", [])
        if step.get("id")
    }


def _describe_target(edit: dict) -> str:
    target = edit.get("target") or {}
    return (
        target.get("stepId")
        or target.get("actionId")
        or target.get("group")
        or "your edit"
    )


def dropped_edit_text(edit: dict) -> str:
    """User-facing activity line for a delta D2 could not replay.

    Deliberately named after WHAT the user touched, not the internal op --
    the same "effect, not mechanism" rule as resolver_progress_text.
    """
    return (
        f"Your manual change to {_describe_target(edit)} no longer applies "
        "after this update, so it was dropped."
    )


def _effective_edits(edits: list[dict] | None) -> list[dict]:
    """Collapse superseded gestures into the final user-visible changes."""
    latest: dict[tuple[str, str], tuple[int, dict]] = {}
    for index, edit in enumerate(edits or []):
        op = str((edit or {}).get("op") or "")
        target = (edit or {}).get("target") or {}
        target_id = (
            target.get("actionId")
            or target.get("group")
            or target.get("stepId")
            or f"unknown:{index}"
        )
        # A task move already includes allocation and ordering. Keep the key
        # per operation so unrelated route/position edits on the same task are
        # still reported separately.
        latest[(op, str(target_id))] = (index, edit)
    return [item for _index, item in sorted(latest.values(), key=lambda pair: pair[0])]


def _action_label(action_id: str | None, actions: list[dict] | None) -> str:
    if not action_id:
        return "the task"
    action = next(
        (item for item in actions or [] if isinstance(item, dict) and item.get("id") == action_id),
        None,
    )
    if action:
        op = action.get("op")
        if op == "move" and action.get("object") and action.get("dest"):
            return f"{action['object']} → {action['dest']}"
        if op == "open" and action.get("facility"):
            return f"open {action['facility']}"
        if op == "close" and action.get("facility"):
            return f"close {action['facility']}"
        if op == "go_to" and action.get("target"):
            return f"go to {action['target']}"
    return action_id


def _task_for_step(plan: dict | None, step_id: str | None) -> str | None:
    if not isinstance(plan, dict) or not step_id:
        return None
    for task in plan.get("tasks", []):
        if any(step.get("id") == step_id for step in task.get("steps", [])):
            return task.get("task")
    return None


def _task_robot(plan: dict | None, action_id: str | None) -> str | None:
    if not isinstance(plan, dict) or not action_id:
        return None
    task = next(
        (item for item in plan.get("tasks", []) if item.get("task") == action_id),
        None,
    )
    return task.get("robot") if task else None


def applied_edit_message(
    edits: list[dict] | None,
    before_plan: dict | None,
    after_plan: dict | None,
    actions: list[dict] | None,
    *,
    dropped_count: int = 0,
) -> str:
    """Describe successfully applied manual edits without an LLM call.

    Facts come only from the replay-success subset supplied by the compound
    turn. Repeated gestures targeting the same field collapse to the final
    effective edit, and task/step ids are resolved through semantic actions
    and the before/after plans.
    """
    lines: list[str] = []
    for edit in _effective_edits(edits):
        op = edit.get("op")
        target = edit.get("target") or {}
        action_id = target.get("actionId") or target.get("group")

        if op == "remove_task":
            lines.append(f"Removed {_action_label(action_id, actions)}.")
            continue

        if op == "move_task":
            label = _action_label(action_id, actions)
            source_robot = _task_robot(before_plan, action_id)
            robot = _task_robot(after_plan, action_id) or edit.get("robot")
            anchor_id = edit.get("afterActionId")
            position = (
                f" after {_action_label(anchor_id, actions)}"
                if anchor_id
                else " at the start of the task sequence"
            )
            if source_robot and robot and source_robot != robot:
                lines.append(f"Moved {label} to {robot}{position}.")
            else:
                suffix = f" on {robot}" if robot else ""
                lines.append(f"Reordered {label}{position}{suffix}.")
            continue

        if op == "set_task_robot":
            robot = _task_robot(after_plan, action_id) or edit.get("robot")
            lines.append(f"Assigned {_action_label(action_id, actions)} to {robot}.")
            continue

        step_id = target.get("stepId")
        task_id = _task_for_step(after_plan, step_id) or _task_for_step(before_plan, step_id)
        label = _action_label(task_id, actions)

        if op == "set_step_after":
            after_ids = edit.get("after") or []
            if after_ids:
                dependency_task = (
                    _task_for_step(after_plan, after_ids[0])
                    or _task_for_step(before_plan, after_ids[0])
                )
                lines.append(
                    f"Made {label} wait until {_action_label(dependency_task, actions)} finishes."
                )
            else:
                lines.append(f"Removed the manual dependency for {label}.")
        elif op == "set_step_at":
            lines.append(f"Updated the position for {label}.")
        elif op == "set_step_via":
            lines.append(f"Updated the route for {label}.")
        elif op == "set_step_standoff":
            lines.append(f"Updated the approach position for {label}.")

    if lines:
        heading = "Applied your edit:" if len(lines) == 1 else f"Applied {len(lines)} edits:"
        message = "\n".join([heading, *[f"{index}. {line}" for index, line in enumerate(lines, 1)]])
    else:
        message = "No edits were applied."
    if dropped_count:
        plural = "s" if dropped_count != 1 else ""
        message += f"\n{dropped_count} edit{plural} could not be applied."
    return message


def replay_edits(plan: dict, edits: list[dict] | None) -> tuple[dict, list[dict]]:
    """Apply each delta in order onto a copy of ``plan``.

    Returns ``(new_plan, dropped)`` -- ``dropped`` is every delta whose
    target id was not found in ``plan`` (see module docstring). Unknown `op`
    values are also dropped rather than raising: a delta from a newer
    frontend build talking to an older backend must degrade safely, not
    break the turn.
    """
    if not edits:
        return plan, []
    result = copy.deepcopy(plan)
    dropped: list[dict] = []
    for edit in edits:
        op = edit.get("op")
        target = edit.get("target") or {}
        applied = False

        if op == "move_task":
            action_id = target.get("actionId")
            robot = edit.get("robot")
            after_action_id = edit.get("afterActionId")
            tasks = result.get("tasks", [])
            source_index = next(
                (index for index, task in enumerate(tasks)
                 if task.get("task") == action_id), None)
            anchor_index = (
                next((index for index, task in enumerate(tasks)
                      if task.get("task") == after_action_id), None)
                if after_action_id is not None else None
            )
            anchor_valid = (
                after_action_id is None
                or (anchor_index is not None
                    and after_action_id != action_id
                    and tasks[anchor_index].get("robot") == robot)
            )
            if source_index is not None and anchor_valid:
                task = tasks.pop(source_index)
                if task.get("robot") != robot:
                    task["robot"] = robot
                    task["robot_locked"] = True
                if after_action_id is not None:
                    insert_at = next(
                        index for index, item in enumerate(tasks)
                        if item.get("task") == after_action_id
                    ) + 1
                else:
                    insert_at = next(
                        (index for index, item in enumerate(tasks)
                         if item.get("robot") == robot), len(tasks))
                tasks.insert(insert_at, task)
                applied = True

        elif op == "set_task_robot":
            group = target.get("actionId") or target.get("group")
            robot = edit.get("robot")
            for task in result.get("tasks", []):
                if task.get("task") == group:
                    task["robot"] = robot
                    task["robot_locked"] = True
                    applied = True
                    break

        elif op == "set_step_after":
            step = _find_step(result, target.get("stepId"))
            if step is not None:
                promote_edited_artifact(step, op)
                known = _known_step_ids(result)
                after_ids = [
                    a for a in dict.fromkeys(edit.get("after") or [])
                    if a != step.get("id") and a in known
                ]
                if after_ids:
                    step["after"] = after_ids
                else:
                    step.pop("after", None)
                applied = True

        elif op == "set_step_at":
            step = _find_step(result, target.get("stepId"))
            if step is not None:
                promote_edited_artifact(step, op)
                at = edit.get("at")
                if at is not None:
                    step["at"] = list(at)[:2]
                    applied = True

        elif op == "set_step_via":
            step = _find_step(result, target.get("stepId"))
            if step is not None:
                promote_edited_artifact(step, op)
                via = edit.get("via")
                if via:
                    step["via_points"] = [list(point) for point in via]
                else:
                    step.pop("via_points", None)
                    step.pop("route", None)
                applied = True

        elif op == "set_step_standoff":
            # Not one of D2's four enumerated ops, but every planEdits.* call
            # site must push a delta (item 16) and setStepStandoff is a real
            # call site (ScenePage's navigate-standoff drag). No resolver
            # tool moves a navigate's standoff, so this op carries no
            # protected-set entry -- it exists purely so the edit replays
            # instead of being silently lost on the next author turn.
            step = _find_step(result, target.get("stepId"))
            if step is not None:
                promote_edited_artifact(step, op)
                standoff = edit.get("standoff")
                if standoff:
                    step["standoff"] = list(standoff)[:2]
                else:
                    step.pop("standoff", None)
                applied = True

        if not applied:
            dropped.append(edit)

    return result, dropped

File: frontend/src/plan/planEdits.ts
/**
 * Pure, immutable edits on an AuthoredPlan. The plan is the single source of
 * truth; every edit returns a NEW plan (never mutates), so it's trivially
 * unit-testable and drives a recompile via usePlanCompile. P1 supports exactly
 * two user edits — reassign a step's robot, and add/remove an `after` dep —
 * plus id assignment on load.
 *
 * Step identity: `after` references and Gantt-bar ↔ step mapping need stable
 * ids, so we assign them up front and preserve any the author/LLM already gave.
 */
import type { AuthoredPlan, AuthoredStep, RobotName } from "./planTypes";

/** All steps in document order, tagged with the robot of their owning task. */
export type FlatStep = AuthoredStep & { robot: RobotName; taskLabel?: string };

/**
 * compound_turn_integration_spec.md §3/D2b: a recorded manual-edit mutation.
 * ScenePage keeps an ordered `edits: PlanEditDelta[]` of PENDING (not yet
 * applied) deltas and sends the whole pending batch with the turn that
 * applies it (an edit-tail sync press, or a chat/append turn per D2b's
 * table). The backend applies them onto whichever plan that turn landed on
 * (`plan_edits.replay_edits`, the Python mirror of this module) -- for an
 * "edit" turn or a D1b pure-append graft, that plan is (or extends) the
 * previous turn's RESOLVED plan, not a fresh `decompose(semanticActions)`:
 * D2's original design rebuilt from scratch every time, which meant a delta
 * targeting a resolver-inserted step (a `go_to_rest` departure leg) had no
 * target and silently vanished. `target` is keyed by whatever survives that:
 * a task-level edit targets the task/action id (`actionId`), a step-level
 * edit falls back to the step id (still a real risk for a batch containing
 * its own stale/contradictory edits -- D2's stated risk, spec §3).
 *
 * `set_step_standoff` is NOT one of the four ops the spec's §3 wire shape
 * enumerates, but `setStepStandoff` is a real `planEdits.*` call site
 * (ScenePage's navigate-standoff drag) and item 16 requires EVERY call site
 * to push a delta so it survives an author turn. It carries no protected-set
 * entry in `buildProtectedSet` below (no resolver tool moves a navigate's
 * standoff), so this is purely a replay-fidelity addition, not a pin source.
 */
export type EditTarget = { actionId?: string; group?: string; stepId?: string };

export type PlanEditDelta =
  | { op: "remove_task"; target: { actionId: string } }
  | { op: "move_task"; target: { actionId: string }; robot: RobotName; afterActionId: string | null }
  | { op: "set_task_robot"; target: EditTarget; robot: RobotName }
  | { op: "set_step_after"; target: EditTarget; after: string[] }
  | { op: "set_step_at"; target: EditTarget; at: [number, number] }
  | { op: "set_step_via"; target: EditTarget; via: number[][] }
  | { op: "set_step_standoff"; target: EditTarget; standoff: [number, number] | null };

/** Wire shape of the resolver's manual-edit protected set (integration spec
 *  §3, `body.protected`). Derived from `edits` (D2: "one structure, two
 *  uses") — never stored separately, always recomputed from the current
 *  edits list + the plan those edits' step ids resolve against. */
export type ProtectedSet = {
  allocations: { group: string; robot: RobotName }[];
  orderings: { before: string; after: string }[];
  /** Exact user-authored step edge. Never inferred from compiler sequencing. */
  exact_after_edges?: { step: string; after_step: string }[];
  /** Complete user-authored `after` field for an edited step, including []. */
  exact_after_fields?: { step: string; after: string[] }[];
  destinations: { step: string }[];
  waypoints: { step: string }[];
};

const EMPTY_PROTECTED_SET: ProtectedSet = {
  allocations: [],
  orderings: [],
  exact_after_edges: [],
  exact_after_fields: [],
  destinations: [],
  waypoints: [],
};

/**
 * Derive the resolver's protected set from the accumulated edit deltas
 * (D2/item 16, "one structure, two uses" — the other use is the backend's
 * `replay_edits`). `plan` supplies the step-id -> task/group lookup needed to
 * turn a step-local `set_step_after` edit into both a TASK-level ordering pin
 * for cheap candidate pruning and an exact step edge for preservation. Pass
 * the plan the edit was made against (`livePlan` is fine — group membership
 * does not change under a manual edit).
 */
export function buildProtectedSet(
  edits: PlanEditDelta[],
  plan: AuthoredPlan | null,
): ProtectedSet {
  if (!edits.length) return EMPTY_PROTECTED_SET;
  const groupOf = new Map<string, string>();
  const ephemeralDetourIds = new Set<string>();
  if (plan) {
    for (const task of plan.tasks) {
      for (const step of task.steps) {
        if (step.id) {
          groupOf.set(step.id, (step.group as string | undefined) ?? task.task ?? step.id);
          if (step.compiler_v2_repair === "detour"
              || step.compiler_v2_repair === "go_away") {
            ephemeralDetourIds.add(step.id);
          }
        }
      }
    }
  }
  // `set_step_after` replaces a whole list. The protected representation must
  // match the final replayed edit state, rather than pinning superseded edges
  // from earlier drags in the same pending batch.
  const finalAfterEdit = new Map<string, PlanEditDelta>();
  const finalMoveEdit = new Map<string, Extract<PlanEditDelta, { op: "move_task" }>>();
  for (const edit of edits) {
    if (edit.op === "move_task") {
      finalMoveEdit.set(edit.target.actionId, edit);
      continue;
    }
    if (edit.op !== "set_step_after") continue;
    const key = edit.target.stepId ? `step:${edit.target.stepId}`
      : edit.target.group ? `group:${edit.target.group}` : null;
    if (key) finalAfterEdit.set(key, edit);
  }
  const out: ProtectedSet = {
    allocations: [], orderings: [], exact_after_edges: [], exact_after_fields: [],
    destinations: [], waypoints: [],
  };
  for (const edit of edits) {
    switch (edit.op) {
      case "remove_task":
        // Removal is applied to semantic actions before decomposition. There
        // is no surviving allocation/order/spatial field to protect.
        break;
      case "move_task": {
        if (finalMoveEdit.get(edit.target.actionId) !== edit) break;
        const original = plan?.tasks.find((task) => task.task === edit.target.actionId);
        if (!original) break;
        if (original.robot !== edit.robot) {
          out.allocations.push({ group: edit.target.actionId, robot: edit.robot });
        }
        if (edit.afterActionId && edit.afterActionId !== edit.target.actionId) {
          out.orderings.push({ before: edit.afterActionId, after: edit.target.actionId });
        }
        break;
      }
      case "set_task_robot": {
        const group = edit.target.actionId ?? edit.target.group;
        if (group) out.allocations.push({ group, robot: edit.robot });
        break;
      }
      case "set_step_after": {
        const targetKey = edit.target.stepId ? `step:${edit.target.stepId}`
          : edit.target.group ? `group:${edit.target.group}` : null;
        if (targetKey && finalAfterEdit.get(targetKey) !== edit) break;
        const stepId = edit.target.stepId;
        const stepGroup = stepId ? groupOf.get(stepId) : edit.target.group;
        if (!stepGroup) break;
        if (stepId) {
          out.exact_after_fields!.push({ step: stepId, after: [...edit.after] });
        }
        for (const afterId of edit.after) {
          const afterGroup = groupOf.get(afterId);
          // Only concrete edges populate the legacy edge pins. The complete
          // field pin above also protects an explicitly cleared [] value;
          // compiler implicit sequencing is never promoted into either form.
          if (stepId && afterId && afterId !== stepId && groupOf.has(afterId)) {
            out.exact_after_edges!.push({ step: stepId, after_step: afterId });
          }
          // The edit means "stepGroup waits for afterGroup" -> the intended
          // order is afterGroup BEFORE stepGroup.
          if (afterGroup && afterGroup !== stepGroup) {
            out.orderings.push({ before: afterGroup, after: stepGroup });
          }
        }
        break;
      }
      case "set_step_at":
        if (edit.target.stepId) out.destinations.push({ step: edit.target.stepId });
        break;
      case "set_step_via":
        if (edit.target.stepId && !ephemeralDetourIds.has(edit.target.stepId)) {
          out.waypoints.push({ step: edit.target.stepId });
        }
        break;
      case "set_step_standoff":
        // No resolver tool moves a navigate's standoff today -- nothing to pin.
        break;
    }
  }
  return out;
}

let idCounter = 0;

/**
 * Ensure every step has a stable `id`. Existing ids are kept (so `after` refs
 * and prior edits survive); missing ones get a fresh `s{n}`. Returns a new plan.
 */
export function ensureStepIds(plan: AuthoredPlan): AuthoredPlan {
  const seen = new Set<string>();
  for (const task of plan.tasks) {
    for (const step of task.steps) {
      if (step.id) seen.add(step.id);
    }
  }
  const freshId = () => {
    let id = `s${idCounter++}`;
    while (seen.has(id)) id = `s${idCounter++}`;
    seen.add(id);
    return id;
  };
  return {
    ...plan,
    tasks: plan.tasks.map((task) => ({
      ...task,
      steps: task.steps.map((step) => (step.id ? step : { ...step, id: freshId() })),
    })),
  };
}

/** Locate a step by id. Returns null if not found. */
function findStep(plan: AuthoredPlan, stepId: string): AuthoredStep | null {
  for (const task of plan.tasks) {
    for (const step of task.steps) {
      if (step.id === stepId) return step;
    }
  }
  return null;
}

/**
 * Reassign a step to another robot. Robot is a task-level property, so the step
 * is moved out of its current task into a task owned by `robot`: it joins that
 * robot's most recent existing task, else a new one is created. Relative order
 * is preserved (appended at the end of the destination). No-op (same plan
 * reference) if the step is already on `robot` or the id is unknown.
 */
export function setStepRobot(
  plan: AuthoredPlan,
  stepId: string,
  robot: RobotName,
): AuthoredPlan {
  const step = findStep(plan, stepId);
  if (!step) return plan;

  const owningTask = plan.tasks.find((t) => t.steps.some((s) => s.id === stepId));
  if (owningTask && owningTask.robot === robot && owningTask.steps.length === 1) {
    return plan; // already solely on the target robot — nothing to move
  }

  // Remove the step from its current task (dropping now-empty tasks).
  const withoutStep = plan.tasks
    .map((t) => ({ ...t, steps: t.steps.filter((s) => s.id !== stepId) }))
    .filter((t) => t.steps.length > 0);

  // Append to the last task already owned by the destination robot, or make one.
  let placed = false;
  const tasks = withoutStep.map((t) => ({ ...t, steps: [...t.steps] }));
  for (let i = tasks.length - 1; i >= 0; i--) {
    if (tasks[i].robot === robot) {
      tasks[i].steps.push(step);
      placed = true;
      break;
    }
  }
  if (!placed) {
    const label = typeof step.group === "string" ? step.group : undefined;
    tasks.push({ task: label, robot, steps: [step] });
  }
  return { ...plan, tasks };
}

/**
 * Reassign the WHOLE task containing `stepId` to `robot` (robot is a semantic,
 * task-level property — a task is one robot's coherent job). Flips that task's
 * `robot`; all its steps move lanes together. No-op if already on `robot` or the
 * id is unknown. Step-level moves are intentionally not offered (use drag for
 * per-step ordering/deps).
 */
export function setTaskRobot(
  plan: AuthoredPlan,
  stepId: string,
  robot: RobotName,
): AuthoredPlan {
  let hit = false;
  const tasks = plan.tasks.map((task) => {
    if (task.robot === robot) return task;
    if (task.steps.some((s) => s.id === stepId)) {
      hit = true;
      return { ...task, robot, robot_locked: true };
    }
    return task;
  });
  return hit ? { ...plan, tasks } : plan;
}

/**
 * First and last AUTHORED step ids of a semantic task, or null when the task is
 * unknown or carries no ids.
 *
 * Task-level dependency edits must resolve through this rather than through
 * compiled Gantt keys. A task bar's first/last compiled step is frequently
 * compiler-generated (`#reposition`, `#detour_`, `#go_to_rest`, `#yield`) and
 * exists in no authored plan, so `setStepAfter` would drop it and return the
 * plan unchanged — an edit that looked queued but meant nothing.
 */
export function taskStepBounds(
  plan: AuthoredPlan,
  actionId: string,
): { first: string; last: string } | null {
  const task = plan.tasks.find((candidate) => candidate.task === actionId);
  const ids = (task?.steps ?? [])
    .map((step) => step.id)
    .filter((id): id is string => typeof id === "string" && id.length > 0);
  if (ids.length === 0) return null;
  return { first: ids[0], last: ids[ids.length - 1] };
}

/**
 * Move one complete semantic task into a robot's program order. `afterActionId`
 * is another task on the destination robot; null means the first slot in that
 * robot's lane. Same-robot moves only reorder. Cross-robot moves also pin the
 * allocation so automated repair cannot hand the task back.
 */
export function moveTask(
  plan: AuthoredPlan,
  actionId: string,
  robot: RobotName,
  afterActionId: string | null,
): AuthoredPlan {
  const sourceIndex = plan.tasks.findIndex((task) => task.task === actionId);
  if (sourceIndex < 0 || afterActionId === actionId) return plan;

  const source = plan.tasks[sourceIndex];
  const anchor = afterActionId
    ? plan.tasks.find((task) => task.task === afterActionId)
    : null;
  if (afterActionId && (!anchor || anchor.robot !== robot)) return plan;

  const remaining = plan.tasks.filter((_, index) => index !== sourceIndex);
  const moved = source.robot === robot
    ? source
    : { ...source, robot, robot_locked: true };
  let insertAt: number;
  if (afterActionId) {
    insertAt = remaining.findIndex((task) => task.task === afterActionId) + 1;
  } else {
    const firstOnRobot = remaining.findIndex((task) => task.robot === robot);
    insertAt = firstOnRobot < 0 ? remaining.length : firstOnRobot;
  }

  const tasks = [...remaining.slice(0, insertAt), moved, ...remaining.slice(insertAt)];
  const unchanged = tasks.length === plan.tasks.length
    && tasks.every((task, index) => task === plan.tasks[index]);
  return unchanged ? plan : { ...plan, tasks };
}

/** Add `afterId` to a step's `after` list (deduped). No-op on unknown/self/dupe. */
export function addAfter(
  plan: AuthoredPlan,
  stepId: string,
  afterId: string,
): AuthoredPlan {
  if (stepId === afterId) return plan;
  const step = findStep(plan, stepId);
  if (!step || !findStep(plan, afterId)) return plan;
  if ((step.after ?? []).includes(afterId)) return plan;
  return mapStep(plan, stepId, (s) => ({ ...s, after: [...(s.after ?? []), afterId] }));
}

/**
 * Replace a step's whole `after` list (deduped, self-refs dropped). The empty
 * array clears the key. Used by drag-to-align, which sets one dependency at a
 * time. No-op on unknown id or when the list is unchanged.
 */
export function setStepAfter(
  plan: AuthoredPlan,
  stepId: string,
  afterIds: string[],
): AuthoredPlan {
  const step = findStep(plan, stepId);
  if (!step) return plan;
  const known = new Set<string>();
  for (const task of plan.tasks) for (const s of task.steps) if (s.id) known.add(s.id);
  const next = [...new Set(afterIds)].filter((a) => a !== stepId && known.has(a));
  const cur = step.after ?? [];
  if (cur.length === next.length && cur.every((a, i) => a === next[i])) return plan;
  return mapStep(plan, stepId, (s) => {
    const out = { ...s };
    if (next.length) out.after = next;
    else delete out.after;
    return out;
  });
}

/**
 * Set a place step's explicit drop point `at` (world XY). The backend recomputes
 * the surface z, so only the 2D point is authored — this is the round-trip hook
 * for dragging the place marker in the 3D scene. Rounded to mm to avoid float
 * noise churning recompiles. No-op on unknown id or an unchanged point.
 */
export function setStepAt(
  plan: AuthoredPlan,
  stepId: string,
  at: [number, number],
): AuthoredPlan {
  const step = findStep(plan, stepId);
  if (!step) return plan;
  const next: [number, number] = [
    Math.round(at[0] * 1000) / 1000,
    Math.round(at[1] * 1000) / 1000,
  ];
  const cur = step.at;
  if (cur && cur[0] === next[0] && cur[1] === next[1]) return plan;
  return mapStep(plan, stepId, (s) => ({ ...s, at: next }));
}

/**
 * Set a navigate step's base standoff (world XY where the robot dwells). The
 * backend keeps the facing (aimed at the target) and only moves the dwell point;
 * null clears the override so the default standoff is used again. Only meaningful
 * for a navigate that isn't feeding an articulation replay (the backend ignores
 * it there). Rounded to mm. No-op on unknown id or an unchanged point.
 */
export function setStepStandoff(
  plan: AuthoredPlan,
  stepId: string,
  xy: [number, number] | null,
): AuthoredPlan {
  const step = findStep(plan, stepId);
  if (!step) return plan;
  const next = xy ? ([Math.round(xy[0] * 1000) / 1000, Math.round(xy[1] * 1000) / 1000] as [number, number]) : null;
  const cur = (Array.isArray(step.standoff) ? (step.standoff as [number, number]) : null) ?? null;
  const same = (cur === null && next === null) || (!!cur && !!next && cur[0] === next[0] && cur[1] === next[1]);
  if (same) return plan;
  return mapStep(plan, stepId, (s) => {
    const out = { ...s };
    if (next) out.standoff = next;
    else delete out.standoff;
    return out;
  });
}

/**
 * Set a navigate step's authored `via_points` constraints. The backend plans
 * and returns a separate full `route`; an empty/null value restores automatic
 * routing. Rounded to mm to avoid float-churn recompiles.
 */
export function setStepViaPoints(
  plan: AuthoredPlan,
  stepId: string,
  route: [number, number][] | null,
): AuthoredPlan {
  const step = findStep(plan, stepId);
  if (!step) return plan;
  const r =
    route && route.length
      ? route.map(([x, y]) => [Math.round(x * 1000) / 1000, Math.round(y * 1000) / 1000] as [number, number])
      : null;
  const cur = (Array.isArray(step.via_points) ? step.via_points : null) ?? null;
  const same =
    (cur === null && r === null) ||
    (!!cur && !!r && cur.length === r.length && cur.every((p, i) => p[0] === r[i][0] && p[1] === r[i][1]));
  if (same) return plan;
  return mapStep(plan, stepId, (s) => {
    const out = { ...s };
    if (r) out.via_points = r;
    else delete out.via_points;
    delete out.route;
    return out;
  });
}

/** Remove `afterId` from a step's `after` list. No-op if absent. */
export function removeAfter(
  plan: AuthoredPlan,
  stepId: string,
  afterId: string,
): AuthoredPlan {
  const step = findStep(plan, stepId);
  if (!step || !(step.after ?? []).includes(afterId)) return plan;
  return mapStep(plan, stepId, (s) => {
    const after = (s.after ?? []).filter((a) => a !== afterId);
    const next = { ...s };
    if (after.length) next.after = after;
    else delete next.after;
    return next;
  });
}

/** Apply `fn` to the step with `stepId`, returning a new plan. */
function mapStep(
  plan: AuthoredPlan,
  stepId: string,
  fn: (step: AuthoredStep) => AuthoredStep,
): AuthoredPlan {
  return {
    ...plan,
    tasks: plan.tasks.map((task) => ({
      ...task,
      steps: task.steps.map((step) => (step.id === stepId ? fn(step) : step)),
    })),
  };
}
